cpp/do_module/do_module.cpp

```diff
@@ -0,0 +1,313 @@
+#include <fmt/core.h>
+#include <boost/algorithm/string/join.hpp>
+#include <boost/algorithm/string/predicate.hpp>
+#include <mgp.hpp>
+
+#include "mgclient.hpp"
+
+const char *kProcedureCase = "case";
+const char *kArgumentConditionals = "conditionals";
+const char *kArgumentElseQuery = "else_query";
+const char *kArgumentParams = "params";
+
+const char *kProcedureWhen = "when";
+const char *kArgumentCondition = "condition";
+const char *kArgumentIfQuery = "if_query";
+
+const char *kReturnValue = "value";
+
+const std::vector<std::string> kGlobalOperations = {"CREATE INDEX ON",
+                                                    "DROP INDEX ON",
+                                                    "CREATE CONSTRAINT ON",
+                                                    "DROP CONSTRAINT ON",
+                                                    "SET GLOBAL TRANSACTION ISOLATION LEVEL",
+                                                    "STORAGE MODE IN_MEMORY_TRANSACTIONAL",
+                                                    "STORAGE MODE IN_MEMORY_ANALYTICAL"};
+
+struct ParamNames {
+  std::vector<std::string> node_names;
+  std::vector<std::string> relationship_names;
+  std::vector<std::string> primitive_names;
+};
+
+struct QueryResults {
+  std::vector<std::string> columns;
+  std::vector<std::vector<mg::Value>> results;
+};
+
+ParamNames ExtractParamNames(const mgp::Map &parameters) {
+  std::vector<std::string> node_names;
+  std::vector<std::string> relationship_names;
+  std::vector<std::string> primitive_names;
+
+  for (const auto &map_item : parameters) {
+    if (map_item.value.IsNode()) {
+      node_names.push_back(std::string(map_item.key));
+    } else if (map_item.value.IsRelationship()) {
+      relationship_names.push_back(std::string(map_item.key));
+    } else {
+      primitive_names.push_back(std::string(map_item.key));
+    }
+  }
+
+  return ParamNames{
+      .node_names = node_names, .relationship_names = relationship_names, .primitive_names = primitive_names};
+}
+
+std::string ConstructQueryPreffix(const ParamNames &names) {
```

Comment:
where exactly do you think we would gain performance benefit?

Reply:
On string creation. That's why I suggested that we can use `istream` to send everything in that stream and then produce one string from that stream.

I can't judge if it is a dumb suggestion from me as I think for some queries where batches are huge we are doing bigger concatenations of strings, and it could be an issue, but on the other hand doubt it will make a difference as actually executing of the query will take the most time, so feel free to skip a suggestion and resolve comment.